skyportal/handlers/api/photometry.py

```diff
@@ -493,323 +566,125 @@ def put(self):
+            .outerjoin(Photometry, condition)
+            .filter(Photometry.id.is_(None))
         )
 
-        proxy = DBSession().execute(pkq)
+        new_photometry_ids = [g[0] for g in new_photometry_query]
 
-        # cache this as list for response
-        ids = [i[0] for i in proxy]
-        df['id'] = ids
-        rows = df.where(pd.notnull(df), None).to_dict('records')
+        id_map = {}
 
-        params = []
-        for packet in rows:
-            if packet["filter"] not in instcache[packet['instrument_id']].filters:
-                raise ValidationError(
-                    f"Instrument {instrument.name} has no filter "
-                    f"{packet['filter']}."
+        duplicated_photometry = (
+            DBSession()
+            .query(values_table.c.pdidx, Photometry)
+            .join(values_table, condition)
+            .options(joinedload(Photometry.groups))
+        ).all()
+
+        for pdidx, duplicate in duplicated_photometry:
+            id_map[pdidx] = duplicate.id
+            duplicate_group_ids = set([g.id for g in duplicate.groups])
+
+            # posting to new groups?
+            if len(set(group_ids) - duplicate_group_ids) > 0:
+                # select old + new groups
+                group_ids_update = set(group_ids).union(duplicate_group_ids)
+                groups = (
+                    DBSession()
+                    .query(Group)
+                    .filter(Group.id.in_(group_ids_update))
+                    .all()
                 )
+                # update the corresponding photometry entry in the db
+                duplicate.groups = groups
 
-            flux = packet.pop('standardized_flux')
-            fluxerr = packet.pop('standardized_fluxerr')
-
-            # reduce the DB size by ~2x
-            keys = ['limiting_mag', 'magsys', 'limiting_mag_nsigma']
-            original_user_data = {key: packet[key] for key in keys if key in packet}
-            if original_user_data == {}:
-                original_user_data = None
-
-            phot = dict(
-                id=packet['id'],
-                original_user_data=original_user_data,
-                upload_id=upload_id,
-                flux=flux,
-                fluxerr=fluxerr,
-                obj_id=packet['obj_id'],
-                altdata=packet['altdata'],
-                instrument_id=packet['instrument_id'],
-                ra_unc=packet['ra_unc'],
-                dec_unc=packet['dec_unc'],
-                mjd=packet['mjd'],
-                filter=packet['filter'],
-                ra=packet['ra'],
-                dec=packet['dec'],
-                origin=packet['origin'],
-                hash=packet["hash"],
-            )
+        # now safely drop the duplicates:
+        new = df.loc[new_photometry_ids]
 
-            params.append(phot)
+        if len(new) > 0:
+            try:
+                ids, _ = self.insert_new_photometry_data(new, instcache, group_ids)
```

Comment:
I feel like there is some unnecessary validation here. The "new" photometry has already been filtered for duplicates, but the self.insert_new_photometry_data() function has it's own set of coerce into VALUES and join to check for duplicates logic that I feel can be skipped in this spot.

Reply:
I'm OK with that, I would feel better about it  if we enabled the REPEATBLE READ isolation level for PUT and POST transactions.

https://www.postgresql.org/docs/9.5/transaction-iso.html

Reply:
> The Repeatable Read isolation level only sees data committed before the transaction began; it never sees either uncommitted data or changes committed during transaction execution by concurrent transactions. (However, the query does see the effects of previous updates executed within its own transaction, even though they are not yet committed.) This is a stronger guarantee than is required by the SQL standard for this isolation level, and prevents all of the phenomena described in Table 13-1. As mentioned above, this is specifically allowed by the standard, which only describes the minimum protections each isolation level must provide.

Reply:
Okay, yeah I think REPEATABLE READ makes sense

Reply:
done

Reply:
OK I added a `validate` kwarg to `insert_new_photometry_data` which turns off the validation on request